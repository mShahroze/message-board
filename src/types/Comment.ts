interface User {
  id: string;
  name:string;
}

interface Reply {
  id: string;
  user: User;
  comment: string;
  replies: Reply[];
}

export interface Comment {
  id: string;
  user: User;
  comment: string;
  replies: Reply[];
}

export interface CommentData {
  data: Comment[];
  continuationToken: unknown;
  httpStatusCode: string;
  errors: unknown[];
  isError: boolean;
}

