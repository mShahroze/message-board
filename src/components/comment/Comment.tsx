import React, { useState } from 'react';
import { Comment } from '../../types/Comment';

interface CommentProps {
  comment: Comment;
}

const SingleComment = ({
  comment: {
    id,
    user: { name: username },
    comment,
    replies
  }
}: CommentProps): JSX.Element => {
  const [hidden, setHidden] = useState(true);

  const nestedComments = (replies || []).map((reply) => {
    return <SingleComment key={reply.id} comment={reply} />;
  });

  const handleClick = () => {
    setHidden((s) => !s);
  };

  return (
    <div className='comment'>
      <p>{comment}</p>
      <span>By {username}</span>
      {replies.length && (
        <div
          className='replies replies-dropdown'
          data-testid={`${id}-comment`}
          onClick={() => handleClick()}
        >
          View {replies.length} Replies
        </div>
      )}
      {!hidden && (
        <div data-testid={`${id}-nested-comments`}> {nestedComments} </div>
      )}
    </div>
  );
};

export default SingleComment;
