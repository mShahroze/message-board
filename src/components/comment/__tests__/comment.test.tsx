import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SingleComment from '../Comment';
import { Comment } from '../../../types/Comment';

describe('Comment Component', () => {
  it('should take a snapshot', () => {
    const comment: Comment = {
      id: '01',
      user: { id: '01', name: 'Sky' },
      comment: 'efwefuin fejefeo',
      replies: []
    };
    const { asFragment } = render(<SingleComment comment={comment} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Click on Reply displays more replies for comment', () => {
    const comment: Comment = {
      id: '01',
      user: { id: '01', name: 'Sky' },
      comment: 'Hello my friend!',
      replies: [
        {
          id: '01',
          user: { id: '03', name: 'Kevin' },
          comment: 'My favourite desert is Ice-cream',
          replies: []
        }
      ]
    };
    const { getByTestId } = render(<SingleComment comment={comment} />);
    const viewReply = getByTestId('01-comment');
    fireEvent.click(viewReply);
    const nestedComments = getByTestId('01-nested-comments');
    expect(nestedComments).toBeTruthy();
    expect(nestedComments.children.length).toBe(1);
  });
});
