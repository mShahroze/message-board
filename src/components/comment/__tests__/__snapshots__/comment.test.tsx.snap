// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="comment"
  >
    <p>
      efwefuin fejefeo
    </p>
    <span>
      Sky
    </span>
    0
  </div>
</DocumentFragment>
`;
